File: src/main/scala/GenerateParentheses.scala
import scala.collection.mutable.ListBuffer

object GenerateParentheses extends App {
  private def generate(s: Array[Char], pos:Int,n: Int,open:Int,close:Int, res:ListBuffer[String]):Unit = {
     if(close == n) {
       res += s.mkString("")
     }
     else {
        if(open > close){
          s(pos) = ')'
          generate(s, pos + 1, n, open, close + 1, res)
        }
        if(open < n){
           s(pos) = '('
           generate(s, pos + 1, n, open + 1, close, res)
        }
     }
  }

  def generateParenthesis(n: Int): List[String] = {
     val res = new ListBuffer[String]()
     generate(Array.ofDim[Char](2* n),0, n, 0, 0,res)
     res.toList
  }

  generateParenthesis(3)
}

File: src/main/scala/GenerateParentheses.py
class Solution:
    def generateParenthesis(self, n):
        if not n:
            return []
        res = []

        def paren(open, close, current, res):
            if len(current) == 2*n:
                res.append(current)
                return
            if open < n:
                paren(open+1,close,current+'(', res)
            if open > close:
                paren(open, close+1,current + ')', res)
        paren(0,0, '', res)
        return res


sol = Solution()
print(sol.generateParenthesis(5))